### PR TITLE
Add GBS data interface and an initial dataset

### PR DESCRIPTION
**Pull request by trbromley**
The purpose of this PR is to add the ability for users to easily access pre-generated datasets of samples from GBS generated on some interesting graphs. We also provide a starting dataset.

Subsequent PRs will focus on adding more datasets.

### CODE REVIEW

File: strawberryfields/gbs/data.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+This module provides access to pre-calculated datasets of simulated GBS samples generated from a
+range of input graphs. Each dataset corresponds to a set of samples generated from a fixed graph
+with corresponding values for:
+
+- ``n_mean``: mean number of photons in the GBS device
+- ``n_max``: maximum number of photons allowed in any sample
+-  ``threshold``: flag to indicate whether samples are generated with threshold detection or
+   with photon number resolving detectors.
+- ``n_samples``: total number of samples in dataset
+- ``modes``: number of modes in GBS device or, equivalently, number of nodes in graph
```
Comment (josh146) [inline strawberryfields/gbs/data.py]: Slight nitpick, but I think this would be easier to read with line breaks between bullet points.

File: strawberryfields/gbs/data.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+We use the :class:`Planted` class as an example to show how to interact with the datasets. Datasets
+can be loaded by running:
+
+>>> data = Planted()
+
+Accessing samples from the dataset is then simple:
+
+>>> i = 3
+>>> sample_i = data[i]
+>>> samples = data[:10]
```
Comment (josh146) [inline strawberryfields/gbs/data.py]: I feel the first two lines in this example is a bit confusing, why not just
```rest
>>> sample_3 = data[3]
>>> samples = data[:10]
```

File: strawberryfields/gbs/data.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+Loading data
+^^^^^^^^^^^^
+
+We use the :class:`Planted` class as an example to show how to interact with the datasets. Datasets
+can be loaded by running:
+
+>>> data = Planted()
+
+Accessing samples from the dataset is then simple:
```
Comment (josh146) [inline strawberryfields/gbs/data.py]:
```suggestion
Simply use indexing and slicing to access samples from the dataset:
```

File: strawberryfields/gbs/data.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+>>> sample_i = data[i]
+>>> samples = data[:10]
+
+Datasets also contain metadata relevant to the GBS setup:
+
+>>> data.n_mean
+6
+
+>>> data.n_samples
+50000
```
Comment (josh146) [inline strawberryfields/gbs/data.py]: You may as well define `__len__()`, that way you can do
```rest
>>> len(data)
50000
```
:smile:

Comment (josh146) [reply strawberryfields/gbs/data.py]: (since it is being treated as a Python sequence anyway with indexing, using `len()` feels very natural)

File: strawberryfields/gbs/data.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+Counting photons or clicks in each sample is available using the :meth:`Dataset.counts` method:
+
+>>> data.counts()
+[2, 0, 8, 11, ... , 6]
+
+Code details
+^^^^^^^^^^^^
+
+The :class:`Dataset` class provides the base functionality from which all datasets such as
+:class:`Planted` inherit.
```
Comment (josh146) [inline strawberryfields/gbs/data.py]: no (auto)summary of available datasets?

File: strawberryfields/gbs/data.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+DATA_PATH = pkg_resources.resource_filename("strawberryfields", "gbs/data") + "/"
+
+
+class Dataset(metaclass=ABCMeta):
+    """Base class for loading datasets of pre-generated samples.
+
+    Attributes:
+        n_mean (float): mean number of photons in the GBS device
+        n_max (float): maximum number of photons allowed in any sample. This number is set to
```
Comment (josh146) [inline strawberryfields/gbs/data.py]:
```suggestion
        n_max (float): Maximum number of photons allowed in any sample. This number is set to
```

File: strawberryfields/gbs/data.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+            limit the computation time, any sample being simulated that exceeds ``n_max`` will be
+            ignored and cause calculation to skip to the next sample.
+        threshold (bool): flag to indicate whether samples are generated with threshold detection
+            (i.e., detectors of zero or some photons) or with photon number resolving detectors.
+        adj (array): adjacency matrix of graph from which samples were generated
+        n_samples (int): total number of samples in dataset
+        modes (int): number of modes in GBS device or, equivalently, number of nodes in graph
```
Comment (josh146) [inline strawberryfields/gbs/data.py]: Actually, I think the above can be deleted. It is already provided in both the module docstring _and_ the attribute documentation below, causing it to appear three times in the documentation. It is also at risk of being updated/changed in one place but not another.

File: strawberryfields/gbs/data.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+    """
+
+    _count = 0
+
+    @property
+    @abstractmethod
+    def _data_filename(self) -> str:
+        """Name of files stored in ``./data/``. Samples should be provided as a
```
Comment (josh146) [inline strawberryfields/gbs/data.py]:
```suggestion
        """Base name of files containing the sample data stored in the ``./data/`` directory.
        
        Samples should be provided as a
```

(I like the convention of separating the short summary of the docstring from the remainder of the docstring, I find it makes it easier to skim/parse).

File: strawberryfields/gbs/data.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+        ``scipy.sparse.csr_matrix`` saved in ``.npz`` format and the corresponding adjacency
+        matrix should be provided as a ``.npy`` binary. For ``_data_filename = "example"``,
+        the corresponding samples should be stored as ``./data/example.npz`` and the adjacency
+        matrix as ``./data/example_A.npy``."""
+        pass
+
+    def __init__(self):
+        """Instantiating ``Dataset`` causes the ``scipy.sparse.csr_matrix`` of samples and
+        ``numpy.ndarray`` adjacency matrix to be loaded."""
```
Comment (josh146) [inline strawberryfields/gbs/data.py]: `__init__` doesn't require a docstring in this instance.

File: strawberryfields/gbs/data.py [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+class Planted(Dataset):
+    """Dataset of samples generated from GBS with an embedded 30-node graph containing a dense
+    10-node subgraph planted inside :cite:`arrazola2018using`.
+
+    Attributes:
+        n_mean = 6
+        n_max = 14
+        threshold = True
+        n_samples = 50000
+        modes = 30
```
Comment (josh146) [inline strawberryfields/gbs/data.py]: I feel this should be a bullet point list with backticks

File: strawberryfields/gbs/data.py [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+        if isinstance(key, (slice, tuple)):
+            if isinstance(key, tuple):
+                key = slice(*key)
+            range_tuple = key.indices(self.n_samples)
+            return [self._elem(i) for i in range(*range_tuple)]
+
+        if key < 0:
+            key += self.n_samples
+
+        return self._elem(key)
```
Comment (josh146) [inline strawberryfields/gbs/data.py]: Don't forget to test all these dunder methods thoroughly :laughing: Looks like there could be a lot of edge cases!

File: strawberryfields/gbs/data.py [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+   with photon number resolving detectors.
+- ``n_samples``: total number of samples in dataset
+- ``modes``: number of modes in GBS device or, equivalently, number of nodes in graph
+
+Datasets are available as classes. The following graphs and datasets are provided:
+
+- **Planted graph**: A 30-node graph with a dense 10-node subgraph planted inside
+  :cite:`arrazola2018using`.
+
+  .. image:: ../../_static/planted.png
```
Comment (ixfoduap) [inline strawberryfields/gbs/data.py]: Should probably use our graph visualization module to create this plot :grin:

File: strawberryfields/gbs/data.py (hunk 2 above)
Comment (ixfoduap) [reply strawberryfields/gbs/data.py]: ^ Agree

Comment (ixfoduap) [reply strawberryfields/gbs/data.py]: Also, this is super cool!

File: strawberryfields/gbs/data.py [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+
+>>> data.n_mean
+6
+
+>>> data.n_samples
+50000
+
+Counting photons or clicks in each sample is available using the :meth:`Dataset.counts` method:
+
+>>> data.counts()
```
Comment (ixfoduap) [inline strawberryfields/gbs/data.py]: I don't understand this, what is the output of `counts`? Is it the aggregate of counts over all samples in the dataset? For instance, what does `11` refer to in the output of the example usage?

File: strawberryfields/gbs/data.py (hunk 11 above)
Comment (ixfoduap) [reply strawberryfields/gbs/data.py]: I don't understand dunders :sob: :sob: :sob:

File: strawberryfields/gbs/data.py [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+This module provides access to pre-calculated datasets of simulated GBS samples generated from a
+range of input graphs. Each dataset corresponds to a set of samples generated from a fixed graph
+with corresponding values for:
+
+- ``n_mean``: mean number of photons in the GBS device
+
+- ``n_max``: maximum number of photons allowed in any sample
+
+-  ``threshold``: flag to indicate whether samples are generated with threshold detection or
+   with photon number resolving detectors.
```
Comment (ixfoduap) [inline strawberryfields/gbs/data.py]:
```suggestion
   with photon-number-resolving detectors.
```

File: strawberryfields/gbs/data.py [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+
+- ``n_samples``: total number of samples in dataset
```
Comment (ixfoduap) [inline strawberryfields/gbs/data.py]:
```suggestion
- ``n_samples``: total number of samples in the dataset
```

File: strawberryfields/gbs/data.py [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+
+- ``modes``: number of modes in GBS device or, equivalently, number of nodes in graph
```
Comment (ixfoduap) [inline strawberryfields/gbs/data.py]:
```suggestion
- ``modes``: number of modes in the GBS device or, equivalently, number of nodes in the graph
```

File: strawberryfields/gbs/data.py [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+6
+
+>>> len(data)
+50000
+
+The number of photons or clicks in each sample is available using the :meth:`Dataset.counts` method:
+
+>>> data.counts()
+[2, 0, 8, 11, ... , 6]
+
```
Comment (ixfoduap) [inline strawberryfields/gbs/data.py]: Maybe add a sentence here saying something like: "For example, in this case the fourth sample has 11 photons"?

File: strawberryfields/gbs/data.py [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+DATA_PATH = pkg_resources.resource_filename("strawberryfields", "gbs/data") + "/"
+
+
+class Dataset(metaclass=ABCMeta):
+    """Base class for loading datasets of pre-generated samples.
+
+    Attributes:
+        n_mean (float): mean number of photons in the GBS device
+        n_max (float): Maximum number of photons allowed in any sample. This number is set to
+            limit the computation time, any sample being simulated that exceeds ``n_max`` will be
```
Comment (ixfoduap) [inline strawberryfields/gbs/data.py]:
```suggestion
            limit the computation time: any sample being simulated that exceeds ``n_max`` will be
```

File: strawberryfields/gbs/data.py [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+            ignored and cause calculation to skip to the next sample.
```
Comment (ixfoduap) [inline strawberryfields/gbs/data.py]:
```suggestion
            ignored and cause the calculation to skip to the next sample.
```

File: strawberryfields/gbs/data.py [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+        threshold (bool): flag to indicate whether samples are generated with threshold detection
+            (i.e., detectors of zero or some photons) or with photon number resolving detectors.
```
Comment (ixfoduap) [inline strawberryfields/gbs/data.py]:
```suggestion
            (i.e., detectors of zero or some photons) or with photon-number-resolving detectors.
```

File: strawberryfields/gbs/data.py [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+        adj (array): adjacency matrix of graph from which samples were generated
```
Comment (ixfoduap) [inline strawberryfields/gbs/data.py]:
```suggestion
        adj (array): adjacency matrix of the graph from which samples were generated
```

File: strawberryfields/gbs/data.py [hunk 22]
```diff
@@ -0,0 +1,209 @@
...
+        n_samples (int): total number of samples in dataset
```
Comment (ixfoduap) [inline strawberryfields/gbs/data.py]:
```suggestion
        n_samples (int): total number of samples in the dataset
```

File: strawberryfields/gbs/data.py [hunk 23]
```diff
@@ -0,0 +1,209 @@
...
+        modes (int): number of modes in GBS device or, equivalently, number of nodes in graph
```
Comment (ixfoduap) [inline strawberryfields/gbs/data.py]:
```suggestion
        modes (int): number of modes in the GBS device or, equivalently, number of nodes in graph
```

File: strawberryfields/gbs/data.py [hunk 24]
```diff
@@ -0,0 +1,209 @@
...
+
+    def __init__(self):
+        self.data = scipy.sparse.load_npz(DATA_PATH + self._data_filename + ".npz")
+        self.adj = np.load(DATA_PATH + self._data_filename + "_A.npy")
+        self.n_samples, self.modes = self.data.shape
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
```
Comment (ixfoduap) [inline strawberryfields/gbs/data.py]: For my learning process: Dunders don't need docstrings?

File: strawberryfields/gbs/data.py (hunk 10 above)
Comment (ixfoduap) [reply strawberryfields/gbs/data.py]: I guess Josh's comment is that it would look better with backticks? Would be nice to compare both if it's not too much trouble

File: tests/gbs/test_data.py [hunk 25]
```diff
@@ -0,0 +1,131 @@
...
+        ``axis==1``. The samples are patched to the fixed ``patch_samples`` list."""
+        assert dataset_patched.counts(1) == [0, 1, 1, 2, 1, 2, 2, 3, 5, 5]
+
+    def test_iter(self, dataset_patched):
+        """Test if dataset class allows correct iteration over itself"""
+        assert [i for i in dataset_patched] == self.patch_samples
+
+    def test_slice(self, dataset_patched):
+        """Test if dataset class allows correct slicing over items"""
+        assert [i for i in dataset_patched[1, 4, 2]] == [
```
Comment (ixfoduap) [inline tests/gbs/test_data.py]: Is it normal to spread `[self.patch_samples[1], self.patch_samples[3], ]` over so many lines?

File: strawberryfields/gbs/data.py [hunk 26]
```diff
@@ -0,0 +1,209 @@
...
+GBS Datasets
+============
+
+**Module name:** :mod:`strawberryfields.gbs.data`
+
+.. currentmodule:: strawberryfields.gbs.data
+
+This module provides access to pre-calculated datasets of simulated GBS samples generated from a
+range of input graphs. Each dataset corresponds to a set of samples generated from a fixed graph
+with corresponding values for:
```
Comment (ixfoduap) [inline strawberryfields/gbs/data.py]: Would it make sense to have a `description` attribute that gives more information about the dataset? For example, playing around with the `Planted` dataset I didn't find information regarding which nodes correspond to the planted subgraph. For other datasets we may want to give similar information, like location of max clique, or what class the MUTAG graphs belong to

File: strawberryfields/gbs/data.py (hunk 17 above)
Comment (trbromley) [reply strawberryfields/gbs/data.py]: Nice! Have added

File: strawberryfields/gbs/data.py (hunk 10 above)
Comment (trbromley) [reply strawberryfields/gbs/data.py]: image

^ Although maybe I need to put the backticks in some other configuration?

File: tests/gbs/test_data.py (hunk 25 above)
Comment (trbromley) [reply tests/gbs/test_data.py]: It was an automatic thing from `black` 🤔

File: strawberryfields/gbs/data.py (hunk 26 above)
Comment (josh146) [reply strawberryfields/gbs/data.py]: I think that's a good approach; it can then be shown via `help(Planted)`. Similarly to how it works with NumPy.

File: strawberryfields/gbs/data.py [hunk 27]
```diff
@@ -0,0 +1,211 @@
...
+from abc import ABCMeta, abstractmethod
+
+import pkg_resources
+import numpy as np
+import scipy
+
+DATA_PATH = pkg_resources.resource_filename("strawberryfields", "gbs/data") + "/"
+
+
+class Dataset(metaclass=ABCMeta):
```
Comment (josh146) [inline strawberryfields/gbs/data.py]: Note that while this was the recommended way of creating an abstract class in Python<3.4, since 3.4, the better approach has been to do
```python
from abc import ABC

class Dataset(ABC):
```

i.e., inheritance over defining the metaclass.

File: strawberryfields/gbs/data.py [hunk 28]
```diff
@@ -0,0 +1,211 @@
...
+            <https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html>`__.
+    """
+
+    _count = 0
+
+    @property
+    @abstractmethod
+    def _data_filename(self) -> str:
+        """Base name of files containing the sample data stored in the ``./data/`` directory.
+        Samples should be provided as a ``scipy.sparse.csr_matrix`` saved in ``.npz`` format and
```
Comment (josh146) [inline strawberryfields/gbs/data.py]:
```suggestion

        Samples should be provided as a ``scipy.sparse.csr_matrix`` saved in ``.npz`` format and
```

File: strawberryfields/gbs/data.py [hunk 29]
```diff
@@ -0,0 +1,211 @@
...
+        the corresponding adjacency matrix should be provided as a ``.npy`` binary. For
```
Comment (josh146) [inline strawberryfields/gbs/data.py]: line break for the example?

File: strawberryfields/gbs/data.py (hunk 10 above)
Comment (josh146) [reply strawberryfields/gbs/data.py]: Ah, I was meaning the formatting in the actual code docstring, not the rendered page. But I didn't know you could write it like that and it would render! In which case I'm happy with it as is. Learnt something knew today 💪

File: strawberryfields/gbs/data.py (hunk 26 above)
Comment (trbromley) [reply strawberryfields/gbs/data.py]: Great!

I added:
"The 10-node planted clique is contained within the final 10 nodes of the graph." to the `Planted` docstring

File: strawberryfields/gbs/data.py (hunk 27 above)
Comment (trbromley) [reply strawberryfields/gbs/data.py]: Nice thanks Josh! Done.

File: strawberryfields/gbs/data.py (hunk 28 above)
Comment (trbromley) [reply strawberryfields/gbs/data.py]: Have done manually. Thanks

File: strawberryfields/gbs/data.py (hunk 29 above)
Comment (trbromley) [reply strawberryfields/gbs/data.py]: Done!

File: strawberryfields/gbs/data.py [hunk 30]
```diff
@@ -0,0 +1,215 @@
...
+        if isinstance(key, (slice, tuple)):
+            if isinstance(key, tuple):
+                key = slice(*key)
+            range_tuple = key.indices(self.n_samples)
+            return [self._elem(i) for i in range(*range_tuple)]
+
+        if key < 0:
+            key += self.n_samples
+
+        return self._elem(key)
```
Comment (zzzeid) [inline strawberryfields/gbs/data.py]: Since `__getitem__` can get a variety of inputs, we should probably check the input here and provide some output to the user if they provide an incorrect input (akin to the error you would get from trying to get an element in a list, etc.)

It also could be more readable to flatten the if statements a little, like this:

```python
    def __getitem__(self, key):
        if not isinstance(key, (slice, tuple, int)):
            raise TypeError("Dataset indices must be integers, slices, or tuples")

        if isinstance(key, int):
            return self._elem(key + self.n_samples)
        
        if isinstance(key, tuple):
            key = slice(*key)

        range_tuple = key.indices(self.n_samples)
        return [self._elem(i) for i in range(*range_tuple)]
```

Comment (trbromley) [reply strawberryfields/gbs/data.py]: image

^^ That's cool!

Comment (trbromley) [reply strawberryfields/gbs/data.py]: And thanks for this 😁 Have added!!

File: strawberryfields/gbs/data.py [hunk 31]
```diff
@@ -0,0 +1,217 @@
...
+- ``n_mean``: mean number of photons in the GBS device
+
+- ``n_max``: maximum number of photons allowed in any sample
+
+-  ``threshold``: flag to indicate whether samples are generated with threshold detection or
+   with photon-number-resolving detectors.
+
+- ``n_samples``: total number of samples in the dataset
+
+- ``modes``: number of modes in the GBS device or, equivalently, number of nodes in the graph
```
Comment (josh146) [inline strawberryfields/gbs/data.py]: Nitpick, but for the above, I would either have full-stops for all bullet point lists, or no full stops (to keep it consistent).

Comment (trbromley) [reply strawberryfields/gbs/data.py]: Fixed 😁
